src/edu/washington/escience/myria/operator/agg/StreamingAggregate.java

```diff
@@ -0,0 +1,275 @@
+    aggregatorStates = AggUtils.allocateAggStates(aggregators);
+  }
+
+  /**
+   * Add aggregate results with previous grouping key to result buffer.
+   *
+   * @throws DbException if any error
+   */
+  private void addToResult() throws DbException {
+    int fromIndex = 0;
+    for (; fromIndex < curGroupKey.numColumns(); ++fromIndex) {
+      TupleUtils.copyValue(curGroupKey, fromIndex, 0, resultBuffer, fromIndex);
+    }
+    for (int agg = 0; agg < aggregators.length; ++agg) {
+      aggregators[agg].getResult(resultBuffer, fromIndex, aggregatorStates[agg]);
+      fromIndex += aggregators[agg].getResultSchema().numColumns();
+    }
+  }
+
+  /**
+   * @return A batch's worth of result tuples from this aggregate.
+   * @throws DbException if there is an error.
+   */
+  private TupleBatch getResultBatch() throws DbException {
+    Preconditions.checkState(getChild().eos(), "cannot extract results from an aggregate until child has reached EOS");
+    if (finalBuffer == null) {
+      finalBuffer = resultBuffer.finalResult();
+      if (resultBuffer.numTuples() == 0) {
+        throw new DbException("0 tuples in result buffer");
+      }
+      resultBuffer = null;
+    }
+    if (finalBuffer.isEmpty()) {
+      return null;
+    } else {
+      return finalBuffer.get(0);
+    }
+  }
+
+  /**
+   * The schema of the aggregate output. Grouping fields first and then aggregate fields.
+   *
+   * @return the resulting schema
+   */
+  @Override
+  protected Schema generateSchema() {
+    Operator child = getChild();
+    if (child == null) {
+      return null;
+    }
+    Schema inputSchema = child.getSchema();
+    if (inputSchema == null) {
+      return null;
+    }
+
+    groupSchema = inputSchema.getSubSchema(gFields);
+
+    /* Build the output schema from the group schema and the aggregates. */
+    final ImmutableList.Builder<Type> aggTypes = ImmutableList.<Type> builder();
+    final ImmutableList.Builder<String> aggNames = ImmutableList.<String> builder();
+
+    try {
+      for (Aggregator agg : AggUtils.allocateAggs(factories, inputSchema)) {
+        Schema curAggSchema = agg.getResultSchema();
+        aggTypes.addAll(curAggSchema.getColumnTypes());
+        aggNames.addAll(curAggSchema.getColumnNames());
+      }
+    } catch (DbException e) {
+      throw new RuntimeException("unable to allocate aggregators to determine output schema", e);
+    }
+    aggSchema = new Schema(aggTypes, aggNames);
+    return Schema.merge(groupSchema, aggSchema);
+  }
+
+  @Override
+  protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbException {
+    Preconditions.checkState(getSchema() != null, "unable to determine schema in init");
+    aggregators = AggUtils.allocateAggs(factories, getChild().getSchema());
+    aggregatorStates = AggUtils.allocateAggStates(aggregators);
+    resultBuffer = new TupleBuffer(getSchema());
+  }
+
+  @Override
+  protected void cleanup() throws DbException {
```

Comment:
What is the rationale for all this `null`-setting? I don't see how it's preventing any memory leaks...